src/ctapipe/image/toymodel.py

```diff
@@ -375,20 +375,88 @@ def pdf(self, x, y):
 class RingGaussian(ImageModel):
     """A shower image consisting of a ring with gaussian radial profile.
 
-    Simplified model for a muon ring.
+    Simplified model for a muon ring. With asymmetry in both the x and y directions.
     """
 
-    def __init__(self, x, y, radius, sigma):
+    def __init__(
+        self,
+        x,
+        y,
+        radius,
+        sigma,
+        asymmetry_magnitude=0,
+        asymmetry_orientation_angle_deg=0 * u.rad,
+    ):
         self.unit = x.unit
         self.x = x
         self.y = y
         self.sigma = sigma
         self.radius = radius
+        self.asymmetry_slope_x = asymmetry_magnitude * np.cos(
+            asymmetry_orientation_angle_deg.to_value(u.rad)
+        )
+        self.asymmetry_slope_y = asymmetry_magnitude * np.sin(
+            asymmetry_orientation_angle_deg.to_value(u.rad)
+        )
         self.dist = norm(
             self.radius.to_value(self.unit), self.sigma.to_value(self.unit)
         )
 
     def pdf(self, x, y):
         """2d probability for photon electrons in the camera plane."""
         r = np.sqrt((x - self.x) ** 2 + (y - self.y) ** 2)
-        return self.dist.pdf(r)
+        uniform_ring_pdf = self.dist.pdf(r)
+        asymmetric_ring_pdf = (
+            uniform_ring_pdf
+            * self.linear_pdf(x, self.x, self.asymmetry_slope_x)
+            * self.linear_pdf(y, self.y, self.asymmetry_slope_y)
+        )
+        return (
+            asymmetric_ring_pdf / np.sum(asymmetric_ring_pdf) * np.sum(uniform_ring_pdf)
+        )
+
+    def linear_pdf(self, var, var_shift, linear_slope=0):
+        """This is a function that generates a probability density
+        function (pdf) with a linear increase/decrease and a given slope in 1D.
+
+        Parameters
+        ----------
+        var : u.Quantity[length, shape(n, )]
+              A random variable with a linearly increasing or decreasing probability
+              density function.
+        linear_slope: linear slope
+        var_shift: Allow the PDF to have a shift.
+
+        Returns
+        -------
+        a pdf shape(n, )
+
+        """
+
+        if linear_slope == 0:
+            return 1
+
+        if len(var) < 2:
+            raise ValueError("The var array is too short.")
+
+        if np.isnan(var).any():
+            raise ValueError("array contains nans.")
+
+        min_var = np.nanmin(var)
+        max_var = np.nanmax(var)
```

Comment:
```suggestion
        max_var = np.max(var)
```

The check on NaNs is done above, you don't need to mask it anymore, and regular min/max are about twice faster in our case.

Reply:
@burmist-git Please implement this suggestion

Reply:
> If a log-like probability density function (PDF) receives a value such as -1, it returns NaN due to log(-1). This is expected behavior, and it is common practice to replace such values with a PDF value of 0.

But this is a PDF, not a log-likelihood and your input hre is a pixel coordinate and that should _never_ be NaN. 

Where did you encounter NaNs? I still think this check is not needed or should raise an error.

Reply:
I can repeat my answer again : Change raise Warning to warnings.warn in the last commit. During the pdf generation one can use method's like inverse function (for ex.) and nan can appear. So the warnings.warn will show this warning and the function handles them in an expected manner.

P.S. It is too long now the discussion. The NaN are not passed I will remove any checks including the raising flag.